### ISSUE TITLE
some questions about data

### ISSUE THREAD

**Issue by 3bobo** [rafellerc/Pytorch-SiamFC#12]
Thank your for this job! it's an excellent work. But I still have some questions. 
First, the label is related to the size of output, and if the size is fixed(such as 33 in your code), the label is fixed for all images, in other words, **_are they sharing the same label ?_** cuz the element of the label is computed by the distance of element to the center, and they are the same for all input.
Second, how to train the net? there are many image sequences in training data, each sequence is tracking different object. so **_does the net can only input two images from same sequence  to train each time?_** or it can use two images regardless they are from different sequences.

**Comment by rafellerc** [comment]
Answering the first question: Yes, once the distance thresholds that define the positive and negative regions are set the labels are identical (as you mentioned, because of the fixed size).
Second question: The input to the network must be two **distinct images** from the **same target/object**. In short, the objective of the training is to 'approach' the representations/embeddings of distinct poses/illuminations etc of any given object with respect to the cross-correlation operation. The perfect network would be invariant (more precisely: equivariant) to any transformation on the image of the target that preserves its identity.

**Comment by rafellerc** [comment]
In order to get two distinct images of the same object you must simply sample two diferent frames (that contain the object) from any given sequence. I hope I answered your question.

**Comment by 3bobo** [comment]
Thank you very much！

**Comment by 3bobo** [comment]
_**how could this model train?**_  if I understand the model and the training process correctly, we can get a score map after `match_corr` whose data is  float, then interpolate the size to original size of input search image, and computer loss between this score map and label, but the label are only 1 or 0, could the score get 0 or 1 precisely after training?

**Comment by rafellerc** [comment]
To answer directly your question, the network produces score maps with probability-like elements, so in general they are between [0,1], and given the definition of the sigmoid function, they are never exactly 0 or 1 (As long as the input is finite the sigmoid only maps to ]0,1[ )

**Comment by 3bobo** [comment]
thanks for your detail answer.

**Comment by 3bobo** [comment]
I understood, the net computes the score of 33 size, the scores are probability of the point belong to center,  and then compute loss between scores and  labels which are all 1 or 0.

but in the model code `training/models.py`, I added some to test:
```
if __name__ == "__main__":
    loss_fn = losses.BCELogit_Loss
    net = SiameseNet(BaselineEmbeddingNet(),corr_map_size=33, stride=4)
    x1 = torch.randn(1,3,127,127)
    x2 = torch.randn(1,3,255,255)
    o = net(x1,x2)
    label = create_BCELogit_loss_label(33, 25, 50,upscale_factor=4)
    label = torch.from_numpy(label)
    label = label.unsqueeze(0)
    print('output : ',o)
    print('label.shape: ',label.shape)
    loss = loss_fn(o, label)
    print('loss: ',loss)
```
but the output o is not all between, such as:
```
output :  tensor([[[[ 0.3467,  0.3645,  0.0527,  ..., -0.5376, -1.0444, -1.0904],
          [ 0.0366,  0.0889,  0.0212,  ..., -0.1391, -0.3706, -1.0236],
          [-0.1137,  0.2311,  0.2556,  ...,  0.2870,  0.4804, -0.4946],
          ...,
          [ 1.0849,  0.9532,  1.2339,  ...,  0.2436, -0.3401, -0.3129],
          [ 0.9031,  1.1192,  1.6322,  ...,  0.1664, -0.7415, -0.8755],
          [ 0.6181,  0.9879,  1.4183,  ...,  0.6592, -0.2840, -0.6887]]]],
       grad_fn=<NativeBatchNormBackward>)
```
there are still some scores higher than 1, was I miss something?

**Comment by rafellerc** [comment]
I don't know if you noticed, but in training/models.py,  in the _SiameseNet_ class, the function that computes the correlation between the embeddings (_match_corr_, shown below) doesn't apply the sigmoid to the score map (The last operation is a batchnorm, also _self.upscale_ is never set to True, since I haven't managed to get a bicubic function with autograd). 
```python
    def match_corr(self, embed_ref, embed_srch):
        """ Matches the two embeddings using the correlation layer. As per usual
        it expects input tensors of the form [B, C, H, W].

        Args:
            embed_ref: (torch.Tensor) The embedding of the reference image, or
                the template of reference (the average of many embeddings for
                example).
            embed_srch: (torch.Tensor) The embedding of the search image.

        Returns:
            match_map: (torch.Tensor) The correlation between
        """
        b, c, h, w = embed_srch.shape
        # Here the correlation layer is implemented using a trick with the
        # conv2d function using groups in order to do the correlation with
        # batch dimension. Basically we concatenate each element of the batch
        # in the channel dimension for the search image (making it
        # [1 x (B.C) x H' x W']) and setting the number of groups to the size of
        # the batch. This grouped convolution/correlation is equivalent to a
        # correlation between the two images, though it is not obvious.
        match_map = F.conv2d(embed_srch.view(1, b * c, h, w),
                             embed_ref, groups=b)
        # Here we reorder the dimensions to get back the batch dimension.
        match_map = match_map.permute(1, 0, 2, 3)
        match_map = self.match_batchnorm(match_map)
        if self.upscale:
            match_map = F.interpolate(match_map, self.upsc_size, mode='bilinear',
                                      align_corners=False)

        return match_map
```
And compare that with the function _make_score_map_ in appSiamFC/producer.py, where you have the line `score_map = sigmoid(score_map)` (No need to pay attention to the whole thing):

```python
    def make_score_map(self, img, mode='sigmoid'):
        """
        """
        img = img/255
        # The offset is inserted so that the final size of the score map matches
        # the search image. To know more see "How to overlay the search img with
        # the score map" in Trello/Report. It is half of the dimension of the
        # Smallest Class Equivalent of the Ref image.
        offset = (((self.ref.shape[0] + 1)//4)*4 - 1)//2
        img_mean = img.mean()
        img_padded = np.pad(img, ((offset, offset), (offset, offset), (0, 0)),
                            mode='constant', constant_values=img_mean)
        img_padded = numpy_to_torch_var(img_padded, device)
        srch_emb = self.net.get_embedding(img_padded)
        score_map = self.net.match_corr(self.ref_emb, srch_emb)
        dimx = score_map.shape[-1]
        dimy = score_map.shape[-2]
        score_map = score_map.view(-1, dimy, dimx)
        if mode == 'sigmoid':
            score_map = sigmoid(score_map)
        elif mode == 'norm':
            score_map = score_map - score_map.min()
            score_map = score_map/score_map.max()
        score_map = score_map.unsqueeze(0)
        # We upscale 4 times, because the total stride of the network is 4
        score_map = F.interpolate(score_map, scale_factor=4, mode='bilinear',
                                  align_corners=False)

        score_map = score_map.cpu()
        score_map = torch_var_to_numpy(score_map)

        return score_map
```
The thing that you noticed is that the output score map in the training version is not composed of probabilities, the elements are actually _Logits_ (https://en.wikipedia.org/wiki/Logit), which 'become' probabilities when transformed by a sigmoid. 

The point of using logits instead of the probabilities during training is that we can use PyTorch's function _BCEWithLogitsLoss_ (Binary Cross entropy with Logits), which is analytically equivalent to a BCE on the probabilities, but uses a clever trick to simplify the equations thus being more numerically stable.

When we need the probabilities, as in the visualization app, we simply pass the logits through the sigmoid function (as shown in _make_score_map_)

**TL;DR** The output you printed are not probabilities, they are logits, from which you get the probabilities by applying the sigmoid function.

**Comment by Leong1230** [comment]
Hi! I am still confused about the label. In labels.py:
`
label[:, :, 0] = dist_map <= pos_thr**2
label[:, :, 1] = (dist_map <= pos_thr**2) | (dist_map > neg_thr**2)
`
 The code produces label with label[:, :, 0] represents the target region and label[:, :, 1] represents the background region. But "(dist_map <= pos_thr**2) | ..." for label[:, :, 1] since make the target and background region overlay, which is not expected, right?

Is there something wrong with the code or my understanding?
Need help, thank you!
